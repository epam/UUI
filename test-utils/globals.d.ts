declare module "jsurl";
declare module "query-string";
