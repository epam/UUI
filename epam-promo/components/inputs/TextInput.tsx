import * as React from 'react';
import { withMods, IEditableDebouncer, IEditableDebouncerOptions } from '@epam/uui-core';
import { TextInput as uuiTextInput, TextInputProps } from '@epam/uui-components';
import * as types from '../types';
import { IHasEditMode, EditMode } from '../types';
import { systemIcons } from '../../icons/icons';
import css from './TextInput.scss';

const defaultSize = '36';
const defaultMode = EditMode.FORM;

export interface TextInputMods extends IHasEditMode {
    size?: types.ControlSize;
}

export function applyTextInputMods(mods: TextInputMods) {
    return [css.root, css['size-' + (mods.size || defaultSize)], css['mode-' + (mods.mode || defaultMode)]];
}

export const TextInput = withMods<TextInputProps, TextInputMods>(uuiTextInput, applyTextInputMods, props => ({
    acceptIcon: systemIcons[props.size || defaultSize].accept,
    cancelIcon: systemIcons[props.size || defaultSize].clear,
    dropdownIcon: systemIcons[props.size || defaultSize].foldingArrow,
}));

export const SearchInput = React.forwardRef<HTMLInputElement, TextInputProps & TextInputMods & IEditableDebouncerOptions>((props, ref) => {
    // analytics events are sending in IEditableDebouncer, so we need to avoid sending events in TextInput
    const { getValueChangeAnalyticsEvent, ...textInputProps } = props;

    return (
        <IEditableDebouncer
            {...props}
            render={iEditable => (
                <TextInput
                    icon={systemIcons[props.size || defaultSize].search}
                    onCancel={!!props.value ? () => iEditable.onValueChange('') : undefined}
                    type="search"
                    inputMode="search"
                    ref={ref}
                    {...textInputProps}
                    {...iEditable}
                />
            )}
        />
    );
});
