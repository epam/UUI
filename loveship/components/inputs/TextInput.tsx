import * as types from '@epam/uui';
import { IEditableDebouncerOptions, withMods } from '@epam/uui-core';
import * as uui from '@epam/uui';
import * as uuiComponents from '@epam/uui-components';
import { systemIcons } from '../icons/icons';
import css from './TextInput.module.scss';

const defaultSize = '36';

export interface TextInputMods extends types.IHasEditMode {
    /**
     * @default '36'
     */
    size?: types.ControlSize | '60';
}

export function applyTextInputMods(mods: TextInputMods) {
    return [css['size-' + (mods.size || defaultSize)]];
}

export type TextInputProps = uuiComponents.TextInputProps & TextInputMods;
export const TextInput = withMods<uuiComponents.TextInputProps, TextInputMods>(uui.TextInput, applyTextInputMods, (props) => ({
    acceptIcon: systemIcons[props.size || defaultSize].accept,
    cancelIcon: systemIcons[props.size || defaultSize].clear,
    dropdownIcon: systemIcons[props.size || defaultSize].foldingArrow,
}));

export type SearchInputProps = uuiComponents.TextInputProps & TextInputMods & IEditableDebouncerOptions;
export const SearchInput = withMods<uuiComponents.TextInputProps, TextInputMods & IEditableDebouncerOptions>(uui.SearchInput, applyTextInputMods, (props) => ({
    acceptIcon: systemIcons[props.size || defaultSize].accept,
    cancelIcon: systemIcons[props.size || defaultSize].clear,
    dropdownIcon: systemIcons[props.size || defaultSize].foldingArrow,
    icon: props.icon ?? systemIcons[props.size || defaultSize].search,
}));
