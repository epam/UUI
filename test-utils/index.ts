export * from './helpers';
export * from './mocks';
