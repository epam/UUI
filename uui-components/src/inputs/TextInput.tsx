import * as React from 'react';
import {
    Icon, uuiMod, uuiElement, uuiMarkers, CX, TextInputCoreProps, cx, useUuiContext,
} from '@epam/uui-core';
import { IconContainer } from '../layout';
import css from './TextInput.module.scss';

const ENTER = 'Enter';
const ESCAPE = 'Escape';

export type IRenderInputProps = React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

export interface TextInputProps extends TextInputCoreProps {
    /** Overrides accept (check) icon */
    acceptIcon?: Icon;
    /** Overrides cancel (cross) icon */
    cancelIcon?: Icon;
    /** Overrides dropdown (chevron) icon */
    dropdownIcon?: Icon;
    /** CSS class(es) to put to the HTML Input element */
    inputCx?: CX;
    /** overrides rendering of HTML Input element  */
    renderInput?: (props: IRenderInputProps) => JSX.Element;
}

export const TextInput = /* @__PURE__ */React.forwardRef<HTMLInputElement, TextInputProps>((props, ref) => {
    const context = useUuiContext();
    const [inFocus, setInFocus] = React.useState<boolean>(false);
    const inputElement = React.useRef<HTMLInputElement>();

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        // Android does not support maxLength
        // https://studysection.com/blog/the-html-maxlength-attribute-is-not-working-as-expected-on-android-phones/
        const targetValue = e.target.value;
        let newValue;
        if (props.maxLength && targetValue.length > props.maxLength) {
            newValue = targetValue.slice(0, props.maxLength);
        } else {
            newValue = targetValue;
        }

        props.onValueChange(newValue);

        if (props.getValueChangeAnalyticsEvent) {
            const event = props.getValueChangeAnalyticsEvent(newValue, props.value);
            context.uuiAnalytics.sendEvent(event);
        }
    };

    const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
        props.onKeyDown?.(e);
        if (e.key === ENTER) props.onAccept?.();
        else if (e.key === ESCAPE) props.onCancel?.();
    };

    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        setInFocus(true);
        props.onFocus?.(event);
    };

    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        setInFocus(false);
        props.onBlur?.(event);
    };

    const handleClick = (e: any) => {
        if (e.target.classList.contains(uuiMarkers.clickable)) {
            return e.preventDefault();
        }
        props.onClick?.(e);
    };

    const handleCancel = () => {
        props.onCancel();
        inputElement.current?.focus();
    };

    const handleWrapperFocus = () => {
        inputElement.current?.focus();
    };

    const getInputProps = () => ({
        type: props.type || 'text',
        className: cx(uuiElement.input, props.inputCx),
        disabled: props.isDisabled,
        placeholder: props.placeholder,
        value: props.value || '',
        readOnly: props.isReadonly,
        onKeyDown: handleKeyDown,
        onChange: handleChange,
        onFocus: handleFocus,
        onBlur: handleBlur,
        autoFocus: props.autoFocus,
        ref: inputElement,
        autoComplete: props.autoComplete,
        name: props.name,
        maxLength: props.maxLength,
        inputMode: props.inputMode,
        tabIndex: (props.tabIndex || inFocus || props.isReadonly || props.isDisabled) ? -1 : 0,
        id: props.id,
        required: props.isRequired,
        'aria-invalid': props.isInvalid,
        'aria-disabled': props.isDisabled,
    });

    const icon = props.icon && <IconContainer icon={ props.icon } onClick={ props.onIconClick } />;
    const showIconsOnAction = props.value && !props.isReadonly && !props.isDisabled;

    return (
        <div
            onClick={ props.onClick && handleClick }
            ref={ ref }
            className={ cx(
                css.container,
                uuiElement.inputBox,
                props.isDisabled && uuiMod.disabled,
                props.isReadonly && uuiMod.readonly,
                props.isInvalid && uuiMod.invalid,
                !props.isReadonly && !props.isDisabled && uuiMarkers.clickable,
                !props.isReadonly && inFocus && uuiMod.focus,
                props.cx,
            ) }
            tabIndex={ -1 }
            onFocus={ handleWrapperFocus }
            { ...props.rawProps }
        >
            {props.iconPosition !== 'right' && icon}
            {props.renderInput ? props.renderInput(getInputProps()) : <input { ...getInputProps() } />}
            {props.onAccept && showIconsOnAction && (
                <IconContainer cx={ cx('uui-icon-accept') } isDisabled={ props.isDisabled } icon={ props.acceptIcon } onClick={ props.onAccept } rawProps={ { role: 'button' } } />
            )}
            {props.onCancel && showIconsOnAction && (
                <IconContainer
                    cx={ cx('uui-icon-cancel', uuiMarkers.clickable) }
                    isDisabled={ props.isDisabled }
                    icon={ props.cancelIcon }
                    onClick={ handleCancel }
                    rawProps={ { role: 'button' } }
                />
            )}
            {props.iconPosition === 'right' && icon}
            {props.isDropdown && (
                <IconContainer cx={ cx((props.isReadonly || props.isDisabled) && css.hidden, css.pointer) } icon={ props.dropdownIcon } flipY={ props.isOpen } />
            )}
        </div>
    );
});
