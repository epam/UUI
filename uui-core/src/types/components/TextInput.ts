import {
    ICanBeReadonly,
    IClickable,
    IDisableable,
    IDropdownToggler,
    IEditable,
    IHasCX,
    IHasIcon,
    IHasPlaceholder,
    IAnalyticableOnChange,
    IHasRawProps,
    ICanFocus,
    IHasTabIndex,
} from '../props';

export interface TextInputCoreProps
    extends IHasCX,
        IClickable,
        IDisableable,
        IEditable<string | undefined>,
        IHasPlaceholder,
        IHasIcon,
        ICanBeReadonly,
        IDropdownToggler,
        IAnalyticableOnChange<string>,
        IHasRawProps<React.HTMLAttributes<HTMLInputElement>>,
        ICanFocus<HTMLInputElement>,
        IHasTabIndex {
    /** Enables cancel (cross) icon, and fires when the icon is clicked */
    onCancel?(): void;
    /** Enables accept (check) icon, and fires when the icon is clicked */
    onAccept?(): void;
    /** keydown event handler to put on the HTML input element */
    onKeyDown?(e?: any): void;
    /** Put focus on the element, when component is mounted */
    autoFocus?: boolean;
    /** Standard 'type' attribute to put on the HTML input element (e.g. 'password') */
    type?: string;
    /** Standard [autocomplete attribute]{@link https://www.w3schools.com/tags/att_input_autocomplete.asp} */
    autoComplete?: string;
    /** Standard [name attribute]{@link https://www.w3schools.com/tags/att_input_name.asp} */
    name?: string;
    /** Maximum input length in characters */
    maxLength?: number;
    /** Standard [inputMode attribute]{@link https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inputmode} */
    inputMode?: React.InputHTMLAttributes<HTMLInputElement>['inputMode'];
    /** HTML ID attribute for input */
    id?: string;
    /** Prefix to render before the input. Can be used to display in-place labels */
    prefix?: React.ReactNode;
    /** Suffix to render after the input. Can be used to display unit or currency */
    suffix?: React.ReactNode;
}
