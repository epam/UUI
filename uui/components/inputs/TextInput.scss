@use  '../../assets/styles/index.scss' as *;

.root {
    fill: var(--uui-text-input-icon);
    border-style: solid;
    box-sizing: border-box;
    padding-left: 5px;
    padding-right: 5px;
    border-radius: var(--uui-text-input-border-radius);

    :global(.uui-input) {
        color: var(--uui-text-input-text);
        font-family: var(--font-regular);
        padding-left: 6px;
        padding-right: 6px;

        &::placeholder {
            color: var(--uui-text-input-placeholder);
        }

        &:placeholder-shown {
            text-overflow: ellipsis;
        }

        &::-moz-placeholder {
            opacity: 1;
        }
    }

    &.mode-form {
        background-color: var(--uui-text-input-bg);
        border-color: var(--uui-text-input-border);

        :global(.-clickable) {
            fill: var(--uui-text-input-icon-clickable);

            &:hover {
                fill: var(--uui-text-input-icon-clickable-hover);
            }
        }

        &:hover {
            border-color: var(--uui-text-input-border-hover);
        }

        &:global(.uui-focus) {
            border-color: var(--uui-text-input-border-active);
        }
    }

    &.mode-cell {
        background-color: transparent;
        border-color: transparent;

        :global(.uui-icon):first-child {
            padding-left: 6px;
        }

        :global(.-clickable) {
            fill: transparent;
        }

        &:hover {
            background-color: var(--uui-text-input-bg);
            border-color: var(--uui-text-input-border);

            :global(.-clickable) {
                fill: var(--uui-text-input-icon-clickable);

                &:hover {
                    fill: var(--uui-text-input-icon-clickable-hover);
                }
            }
        }

        &:global(.uui-focus) {
            background-color: var(--uui-text-input-bg);
            border-color: var(--uui-text-input-border-active);
        }
    }

    &:global(.uui-invalid) {
        &.mode-form, &.mode-cell {
            background-color: var(--uui-text-input-bg);
            border-color: var(--uui-text-input-border-error);
        }
    }

    &:global(.uui-readonly) {
        color: var(--uui-text-input-text);
        fill: var(--uui-text-input-icon);
        pointer-events: none;

        &.mode-form {
            background-color: var(--uui-text-input-bg-readonly);
            border-color: var(--uui-text-input-border-readonly);
        }

        &.mode-cell {
            background-color: transparent;
            border-color: transparent;
        }
    }

    &:global(.uui-disabled) {
        color: var(--uui-text-input-text-disabled);
        fill: var(--uui-text-input-icon-disabled);
        pointer-events: none;

        :global(.uui-input) {
            color: var(--uui-text-input-text-disabled);

            &::placeholder {
                color: var(--uui-text-input-placeholder-disabled);
            }
        }

        &.mode-form {
            background-color: var(--uui-text-input-bg-disabled);
            border-color: var(--uui-text-input-border-disabled);
        }

        &.mode-cell {
            background-color: transparent;
            border-color: transparent;
        }
    }

    //Sizes
    &.size-24 {
        @include control-vertical-layout(24px, 1px);
        min-width: 24px;
    }

    &.size-30 {
        @include control-vertical-layout(30px, 1px);
        min-width: 30px;
    }

    &.size-36 {
        @include control-vertical-layout(36px, 1px);
        min-width: 36px;
    }

    &.size-42 {
        @include control-vertical-layout(42px, 1px);
        min-width: 42px;
    }

    &.size-48 {
        @include control-vertical-layout(48px, 1px);
        min-width: 48px;
    }
}
