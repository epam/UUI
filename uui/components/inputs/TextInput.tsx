import React from 'react';
import { withMods, IEditableDebouncer, IEditableDebouncerOptions } from '@epam/uui-core';
import { TextInput as uuiTextInput, TextInputProps as CoreTextInputProps } from '@epam/uui-components';
import { IHasEditMode, EditMode, ControlSize } from '../types';
import { systemIcons } from '../../icons/icons';
import css from './TextInput.module.scss';

const defaultSize = '36';
const defaultMode = EditMode.FORM;

export interface TextInputMods extends IHasEditMode {
    /**
     * @default '36'
     */
    size?: ControlSize;
}

export function applyTextInputMods(mods: TextInputMods) {
    return [
        css.root,
        css['size-' + (mods.size || defaultSize)],
        css['mode-' + (mods.mode || defaultMode)],
    ];
}

export interface TextInputProps extends CoreTextInputProps, TextInputMods {}
export interface SearchInputProps extends TextInputProps, IEditableDebouncerOptions {}

export const TextInput = withMods<CoreTextInputProps, TextInputMods>(uuiTextInput, applyTextInputMods, (props) => ({
    acceptIcon: systemIcons[props.size || defaultSize].accept,
    cancelIcon: systemIcons[props.size || defaultSize].clear,
    dropdownIcon: systemIcons[props.size || defaultSize].foldingArrow,
}));

export const SearchInput = React.forwardRef<HTMLInputElement, SearchInputProps>((props, ref) => {
    // analytics events are sending in IEditableDebouncer, so we need to avoid sending events in TextInput
    const { ...textInputProps } = props;
    delete textInputProps.getValueChangeAnalyticsEvent;

    return (
        <IEditableDebouncer
            { ...props }
            render={ (iEditable) => (
                <TextInput
                    icon={ systemIcons[props.size || defaultSize].search }
                    onCancel={ !!props.value
                        // In a lot of places, it is required to check if a clicked element is a part of some other element.
                        // Usually, those are global click event handlers. To allow that logic to work correctly, it is necessary
                        // to execute the `disappearing` of the cross (setState execution) after the event will pass through all the handlers.
                        // Related issue: https://github.com/epam/UUI/issues/1045.
                        ? (() => setTimeout(() => iEditable.onValueChange(''), 0))
                        : undefined }
                    type="search"
                    inputMode="search"
                    ref={ ref }
                    { ...textInputProps }
                    { ...iEditable }
                />
            ) }
        />
    );
});
