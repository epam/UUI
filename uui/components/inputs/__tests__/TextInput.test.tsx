import React from 'react';
import { TextInput } from '../TextInput';
import { renderSnapshotWithContextAsync } from '@epam/test-utils';
import { ReactComponent as CalendarIcon } from '../../../icons/calendar-18.svg';

describe('TextInput', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <TextInput
                value={ null }
                onValueChange={ jest.fn }
            />,
        );

        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <TextInput
                value={ null }
                onValueChange={ jest.fn }
                onAccept={ jest.fn }
                onCancel={ jest.fn }
                icon={ CalendarIcon }
                iconPosition="right"
                size="36"
                isDropdown
                isOpen
            />,
        );

        expect(tree).toMatchSnapshot();
    });
});

