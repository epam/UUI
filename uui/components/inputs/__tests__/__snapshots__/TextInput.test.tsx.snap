// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="container uui-input-box root uui-size-36 uui-control-mode-form"
  >
    <input
      class="uui-input"
      tabindex="0"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TextInput should be rendered correctly 2`] = `
<DocumentFragment>
  <div
    class="container uui-input-box root uui-size-36 uui-control-mode-form"
  >
    <input
      class="uui-input"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      class="container uui-icon uui-enabled controlIcon"
    >
      <svg
        class=""
      />
    </div>
    <div
      class="container uui-icon uui-enabled pointer"
    >
      <svg
        class="flipY"
      />
    </div>
  </div>
</DocumentFragment>
`;
